FreeRTOS_UDP_IP.c

```diff
@@ -59,28 +60,129 @@
 
 /*-----------------------------------------------------------*/
 
+static eARPLookupResult_t prvLookupIPInCache( NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                              uint32_t * pulIPAddress );
+
+static void prvSetIPHeaderForICMP( NetworkBufferDescriptor_t * const pxNetworkBuffer );
+
+static void prvSetIPHeaderForUDP( NetworkBufferDescriptor_t * const pxNetworkBuffer );
+
+static eARPLookupResult_t prvStartLookup( NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                          BaseType_t * pxLostBuffer,
+                                          uint32_t ulIPAddress );
+
+static void prvUDPSendPacket( NetworkBufferDescriptor_t * const pxNetworkBuffer );
+
+static void prvFindIPv4Endpoint( NetworkBufferDescriptor_t * const pxNetworkBuffer );
+
+static BaseType_t prvHandleUDPPacketWithoutSocket( NetworkBufferDescriptor_t * pxNetworkBuffer,
+                                                   uint16_t usPort );
+
+/*-----------------------------------------------------------*/
+
+
 /**
- * @brief Process the generated UDP packet and do other checks before sending the
- *        packet such as ARP cache check and address resolution.
+ * @brief Look-up the target IP-address, works for both IPv4 and IPv6.
  *
- * @param[in] pxNetworkBuffer: The network buffer carrying the packet.
+ * @param[in] pxNetworkBuffer: The network buffer carrying the UDP or ICMP packet.
+ * @param[in,out] pulIPAddress: only used for IPv4: the IP-address of the destination.
```

Comment:
How is routed IPv6 handled then?

Reply:
When I wrote the Doxygen comments, I notice the asymmetry: only an IPv4 address is provided! The IPv6 address is passed indirectly, as `pxUDPPacket->xEthernetHeader.xDestinationAddress`.

The IP-address pointed to by `pulIPAddress` may be replaced by the gateway.

It looks like this needs some revision, I will come back to this.